tensorboard/webapp/widgets/line_chart_v2/lib/renderer/threejs_renderer.ts

```diff
@@ -32,13 +37,97 @@ function createOpacityAdjustedColor(hex: string, opacity: number): THREE.Color {
   return new THREE.Color(newD3Color.brighter(1.6 - opacity).hex());
 }
 
+enum CacheType {
+  CIRCLE,
+  LINE,
+  TRIANGLE,
+}
+
 interface LineCacheValue {
-  type: 'line';
-  obj3d: THREE.Object3D;
+  type: CacheType.LINE;
+  obj3d: THREE.Line;
   data: Polyline;
 }
 
-type CacheValue = LineCacheValue;
+interface TriangleCacheValue {
+  type: CacheType.TRIANGLE;
+  obj3d: THREE.Mesh;
+  data: Point;
+}
+
+interface CircleCacheValue {
+  type: CacheType.CIRCLE;
+  obj3d: THREE.Mesh;
+  data: {loc: Point; radius: number};
+}
+
+type CacheValue = LineCacheValue | TriangleCacheValue | CircleCacheValue;
+
+/**
+ * Updates BufferGeometry with Float32Array that denotes flattened array of Vec2 (<x, y>).
+ */
+function updateGeometryWithVec2Array(
+  geometry: THREE.BufferGeometry,
+  flatVec2: Float32Array
+) {
+  const length = flatVec2.length / 2;
+  let positionAttributes = geometry.attributes[
+    'position'
+  ] as THREE.BufferAttribute;
+  if (!positionAttributes || positionAttributes.count !== length * 3) {
+    positionAttributes = new THREE.BufferAttribute(
+      new Float32Array(length * 3),
+      3
+    );
+    geometry.addAttribute('position', positionAttributes);
+  }
+  const values = positionAttributes.array as number[];
+  for (let index = 0; index < length; index++) {
+    values[index * 3] = flatVec2[index * 2];
+    values[index * 3 + 1] = flatVec2[index * 2 + 1];
+  }
+  positionAttributes.needsUpdate = true;
+  geometry.setDrawRange(0, length * 3);
+  // Need to update the bounding sphere so renderer does not skip rendering
+  // this object because it is outside of the camera viewpoint (frustum).
+  geometry.computeBoundingSphere();
+}
+
+/**
+ * Updates an THREE.Object3D like object with geometry and material. Returns true if
+ * geometry is updated with the new data but returns false if we can minimally change the
+ * material to make an update.
+ */
+function updateObject(
+  object: THREE.Mesh | THREE.Line,
+  updateGeometry: (geometry: THREE.BufferGeometry) => THREE.BufferGeometry,
```

Comment:
I'm having a hard time understanding the updateGeometry argument. You don't use the return value. Two of the callers return the input geometry. The third caller returns a brand new geometry without modification to the input.